src/main.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2018 King's College London created by the Software Development
+// Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any
+// person obtaining a copy of this software, associated documentation and/or
+// data (collectively the "Software"), free of charge and under any and all
+// copyright rights in the Software, and any and all patent rights owned or
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and (b) any piece of software and/or hardware listed in the
+// lrgrwrks.txt file if one is included with the Software (each a "Larger Work"
+// to which the Software is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![feature(rustc_private, link_args)]
+extern crate getopts;
+extern crate rustc;
+extern crate rustc_driver;
+extern crate syntax;
+extern crate byteorder;
+extern crate rustc_const_math;
+extern crate mirkat;
+
+use rustc_driver::{driver, CompilerCalls};
+use rustc::ty::TyCtxt;
+use rustc::session::Session;
+use std::process;
+use rustc::hir::def_id::DefId;
+pub use rustc_const_math::ConstInt;
+
+use mirkat::machine::{Machine,Memory};
+
+/// This struct is required by the rustc API in order to implement traits which
+/// are used to hook into the compiler and customise the build pipeline. In
+/// Mirkat, we stop compilation after the MIR has been generated and then hand
+/// the MIR over to the interpreter. This means that Mirkat's invocation of rustc
+/// will never hit LLVM.
+struct MirkatCompilerCalls { }
+
+impl<'a> CompilerCalls<'a> for MirkatCompilerCalls {
```

Comment:
Where in this file is your interpreter invoked?

Reply:
It will be invoked in an entry_point method but I hadn't put this in as part of the current PR. In its current state it would not be entirely correct, as function calls are handled very naively. Would you suggest putting this back in now? Or leaving it for the next PR?

Reply:
It's up to you, but I just thought I might have lost the plot, since I couldn't see how the interpreter was ever entered!

Reply:
Sorry - I should have made this clear. I think it's best I put this in the next PR as function entry is a little outside the scope of this one. I will add an issue though for the sake of transparency.